### PR TITLE
Create a generic container for beat-sorted objects

### LINKED ISSUE #35: Create a generic container for beat-sorted objects
# 📖 Overview

Both BPMs and chart objects are stored the same way and that logic currently exists in the `Chart` class. That logic should be moved out into a generic container class that we can use for both scenarios.

### PR DESCRIPTION
**Pull request by Kangaroux**
Closes #35 

Both BPMs and chart objects are stored:
- In a linked list
- Sorted by beats
- Each element must be on a unique beat

Most of this logic already existed in `Chart` already, it just got moved out into the `BeatObjectList` class. This class implements the full `ICollection` interface and has events for when items are added or removed from the list.

### CODE REVIEW

File: OpenChart.Tests/src/Charting/TestBeatObjectList.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        [Test]
+        public void Test_Add_ObjectsWithSameBeat()
+        {
+            var list = new BeatObjectList<BPM>();
+            list.Add(new BPM(100, 0));
+            Assert.Throws<ArgumentException>(() => list.Add(new BPM(200, 0)));
+        }
+
+        [Test]
+        public void Test_AddMultiple_NullThrowsException()
```
Comment (monotron) [inline OpenChart.Tests/src/Charting/TestBeatObjectList.cs:57]: Please add a test for a list that contains null and non-null values

File: OpenChart/src/Charting/BeatObjectList.cs [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+        /// Adds an object to the list. Multiple object changes cannot occur on the same beat.
+        /// </summary>
+        public void Add(T obj)
+        {
+            if (obj == null)
+            {
+                throw new ArgumentNullException("Object cannot be null.");
+            }
+
+            insertObject(obj);
```
Comment (monotron) [inline OpenChart/src/Charting/BeatObjectList.cs:69]: Rename with PascalCase method name

File: OpenChart/src/Charting/BeatObjectList.cs [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+            obj.CheckValid(objects.Last?.Value, null);
+
+            return objects.AddLast(obj);
+        }
+
+        /// <summary>
+        /// This method is called when an item in the list is modified.
+        /// </summary>
+        private void onItemChanged(object o, EventArgs e)
```
Comment (monotron) [inline OpenChart/src/Charting/BeatObjectList.cs:241]: Please rename with PascalCase
